sql: createPartialStatsPlan failed with 'column a does not have a prior statistic...'

pkg/sql/logictest/tests/5node/5node_test.TestLogic_distsql_stats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7691443?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7691443?buildTab=artifacts#/) on master @ [b5be006bedd7d3cedc3fb3d2248df168e3d64be2](https://github.com/cockroachdb/cockroach/commits/b5be006bedd7d3cedc3fb3d2248df168e3d64be2):


```
=== RUN   TestLogic_distsql_stats
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c1120e87cda9b48ae97a6489658fb611/logTestLogic_distsql_stats3660493154
    test_log_scope.go:79: use -show-logs to present logs inline
    logic.go:2902: let $t_id = 106
    logic.go:2902: let $hist_id_1 = 816923368227799041
    logic.go:2902: let $json_stats = [{"avg_size": 1, "columns": ["a"], "created_at": "2022-11-25 11:27:07.117892", "distinct_count": 4, "histo_buckets": [{"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "1"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "2"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "3"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "4"}], "histo_col_type": "INT8", "histo_version": 2, "name": "s1", "null_count": 0, "row_count": 256}, {"avg_size": 1, "columns": ["b"], "created_at": "2022-11-25 11:27:07.166733", "distinct_count": 4, "histo_buckets": [{"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "1"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "2"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "3"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "4"}], "histo_col_type": "INT8", "histo_version": 2, "null_count": 0, "row_count": 256}]
[11:27:08] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/496/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 48 statements
    logic.go:2902: let $hist_id_3 = 816923373418151941
[11:27:10] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/496/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 102 statements
    logic.go:2902: let $hist_id_1 = 816923379274514437
    logic.go:2902: let $hist_id_1 = 816923379587874821
    logic.go:2902: let $hist_id_1 = 816923379856179205
    logic.go:2902: let $hist_id_1 = 816923380624588805
    logic.go:2902: let $hist_id_1 = 816923384333041669
    logic.go:2902: let $hist_id_1 = 816923384891473925
[11:27:12] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/496/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 159 statements
    logic.go:2902: let $stats = [{"avg_size": 4, "columns": ["x"], "created_at": "2022-11-25 11:27:12.557142", "distinct_count": 1, "histo_buckets": [{"distinct_range": 0, "num_eq": 1, "num_range": 0, "upper_bound": "hi"}], "histo_col_type": "test.public.greeting", "histo_version": 2, "name": "s", "null_count": 0, "row_count": 1}]
    logic.go:2902: let $hist_id_1 = 816923386994327557
    logic.go:2902: let $hist_id_1 = 816923387248705541
    logic.go:2902: let $hist_id_1 = 816923387266990085
[11:27:14] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/496/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 210 statements
    logic.go:2902: let $hist_abcd_a_partial = 816923395317858309
    logic.go:2902: let $hist_abcd_c_partial = 816923395395190789
    logic.go:2902: let $hist_xy_x_partial = 816923395497558021
    logic.go:3182: 
        
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/496/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats:2091: 
        expected success, but found
        (55000) column a does not have a prior statistic, or the prior histogram has no buckets and a partial statistic cannot be collected
        distsql_plan_stats.go:233: in createPartialStatsPlan()
[11:27:15] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/496/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats with config 5node: 243 tests, 1 failures
    logic.go:3968: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/496/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats:2091: error while processing
    logic.go:3968: pq: column a does not have a prior statistic, or the prior histogram has no buckets and a partial statistic cannot be collected
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/c1120e87cda9b48ae97a6489658fb611/logTestLogic_distsql_stats3660493154
--- FAIL: TestLogic_distsql_stats (10.96s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_distsql_stats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-21812

pkg/sql/logictest/tests/5node/5node_test.TestLogic_distsql_stats [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7714725?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7714725?buildTab=artifacts#/) on master @ [1a6e9f885baa124d5ff2996adb966ea15a1a9b2b](https://github.com/cockroachdb/cockroach/commits/1a6e9f885baa124d5ff2996adb966ea15a1a9b2b):


```
=== RUN   TestLogic_distsql_stats
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c1120e87cda9b48ae97a6489658fb611/logTestLogic_distsql_stats720500477
    test_log_scope.go:79: use -show-logs to present logs inline
    logic.go:2902: let $t_id = 106
    logic.go:2902: let $hist_id_1 = 817480316992880641
    logic.go:2902: let $json_stats = [{"avg_size": 1, "columns": ["a"], "created_at": "2022-11-27 10:39:54.391403", "distinct_count": 4, "histo_buckets": [{"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "1"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "2"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "3"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "4"}], "histo_col_type": "INT8", "histo_version": 2, "name": "s1", "null_count": 0, "row_count": 256}, {"avg_size": 1, "columns": ["b"], "created_at": "2022-11-27 10:39:54.497087", "distinct_count": 4, "histo_buckets": [{"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "1"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "2"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "3"}, {"distinct_range": 0, "num_eq": 64, "num_range": 0, "upper_bound": "4"}], "histo_col_type": "INT8", "histo_version": 2, "null_count": 0, "row_count": 256}]
[10:39:55] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1478/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 52 statements
    logic.go:2902: let $hist_id_3 = 817480321768062981
    logic.go:2902: let $hist_id_1 = 817480327187398661
    logic.go:2902: let $hist_id_1 = 817480327521370117
    logic.go:2902: let $hist_id_1 = 817480327767818245
[10:39:57] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1478/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 118 statements
    logic.go:2902: let $hist_id_1 = 817480328929443845
    logic.go:2902: let $hist_id_1 = 817480332202639365
    logic.go:2902: let $hist_id_1 = 817480332700352517
    logic.go:2902: let $stats = [{"avg_size": 4, "columns": ["x"], "created_at": "2022-11-27 10:39:59.577459", "distinct_count": 1, "histo_buckets": [{"distinct_range": 0, "num_eq": 1, "num_range": 0, "upper_bound": "hi"}], "histo_col_type": "test.public.greeting", "histo_version": 2, "name": "s", "null_count": 0, "row_count": 1}]
[10:39:59] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1478/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 183 statements
    logic.go:2902: let $hist_id_1 = 817480334906195973
    logic.go:2902: let $hist_id_1 = 817480335178629125
    logic.go:2902: let $hist_id_1 = 817480335197011973
[10:40:03] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1478/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats: 212 statements
    logic.go:2902: let $hist_abcd_a_partial = 817480347016298501
    logic.go:2902: let $hist_abcd_c_partial = 817480347113848837
    logic.go:2902: let $hist_xy_x_partial = 817480347225325573
    logic.go:3182: 
        
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1478/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats:2091: 
        expected success, but found
        (55000) column a does not have a prior statistic, or the prior histogram has no buckets and a partial statistic cannot be collected
        distsql_plan_stats.go:233: in createPartialStatsPlan()
[10:40:03] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1478/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats with config 5node: 243 tests, 1 failures
    logic.go:3968: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1478/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_stats:2091: error while processing
    logic.go:3968: pq: column a does not have a prior statistic, or the prior histogram has no buckets and a partial statistic cannot be collected
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/c1120e87cda9b48ae97a6489658fb611/logTestLogic_distsql_stats720500477
--- FAIL: TestLogic_distsql_stats (11.77s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_distsql_stats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/alter_table.go
+++ b/pkg/sql/alter_table.go
@@ -1301,6 +1301,11 @@ func insertJSONStatistic(
 	}
 
 	if !settings.Version.IsActive(ctx, clusterversion.V23_1AddPartialStatisticsPredicateCol) {
+
+		if s.PartialPredicate != "" {
+			return pgerror.Newf(pgcode.ObjectNotInPrerequisiteState, "statistic for columns %v with collection time %s to insert is partial but cluster version is below 23.1", s.Columns, s.CreatedAt)
+		}
+
 		_ /* rows */, err := ie.Exec(
 			ctx,
 			"insert-stats",
@@ -1327,6 +1332,12 @@ func insertJSONStatistic(
 			histogram)
 		return err
 	}
+
+	var predicateValue interface{}
+	if s.PartialPredicate != "" {
+		predicateValue = s.PartialPredicate
+	}
+
 	_ /* rows */, err := ie.Exec(
 		ctx,
 		"insert-stats",
@@ -1352,7 +1363,7 @@ func insertJSONStatistic(
 		s.NullCount,
 		s.AvgSize,
 		histogram,
-		s.PartialPredicate,
+		predicateValue,
 	)
 	return err
 }

--- a/pkg/sql/logictest/testdata/logic_test/distsql_stats
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_stats
@@ -1958,11 +1958,6 @@ SELECT * FROM indexed_arr WHERE a @> ARRAY[100]
 
 # Test single column partial statistics creation.
 
-# Disable auto stats collection to avoid new full stats collection
-# when inserting values into the extremes in the table.
-statement ok
-SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false;
-
 statement ok
 CREATE TABLE abcd (a INT PRIMARY KEY, b INT, c INT, d INT, INDEX (c, d));
 
@@ -2085,6 +2080,16 @@ INSERT INTO a_null VALUES (NULL), (1), (2);
 statement ok
 CREATE STATISTICS a_null_stat ON a FROM a_null;
 
+let $a_null_stats
+SHOW STATISTICS USING JSON FOR TABLE a_null
+
+# Inject statistics so that the current stats
+# cache is invalidated and creating partial
+# statistics has access to the latest full
+# statistic.
+statement ok
+ALTER TABLE a_null INJECT STATISTICS '$a_null_stats'
+
 statement ok
 INSERT INTO a_null VALUES (NULL), (NULL), (NULL);
 
@@ -2116,6 +2121,16 @@ INSERT INTO d_desc VALUES (1, 10), (2, 20), (3, 30), (4, 40);
 statement ok
 CREATE STATISTICS sd ON a FROM d_desc;
 
+let $d_desc_stats
+SHOW STATISTICS USING JSON FOR TABLE d_desc;
+
+# Inject statistics so that the current stats
+# cache is invalidated and creating partial
+# statistics has access to the latest full
+# statistic.
+statement ok
+ALTER TABLE d_desc INJECT STATISTICS '$d_desc_stats'
+
 statement ok
 INSERT INTO d_desc VALUES (0, 0), (5, 50);
 
@@ -2145,6 +2160,16 @@ INSERT INTO d_desc VALUES (NULL, NULL), (NULL, 2);
 
 statement ok
 CREATE STATISTICS sdn ON a FROM d_desc;
+
+let $d_desc_null_stats
+SHOW STATISTICS USING JSON FOR TABLE d_desc;
+
+# Inject statistics so that the current stats
+# cache is invalidated and creating partial
+# statistics has access to the latest full
+# statistic.
+statement ok
+ALTER TABLE d_desc INJECT STATISTICS '$d_desc_null_stats'
 
 statement ok
 CREATE STATISTICS sdnp ON a FROM d_desc USING EXTREMES;
@@ -2176,13 +2201,26 @@ statement error pq: table xy does not contain a non-partial forward index with y
 CREATE STATISTICS xy_y_partial ON y FROM xy USING EXTREMES;
 
 statement ok
-DELETE FROM a_null WHERE a IS NOT NULL;
+CREATE TABLE only_null (a INT, INDEX (a));
 
 statement ok
-CREATE STATISTICS a_null_stat ON a FROM a_null;
+INSERT INTO only_null VALUES (NULL), (NULL), (NULL);
+
+statement ok
+CREATE STATISTICS only_null_stat ON a FROM only_null;
+
+let $only_null_stat
+SHOW STATISTICS USING JSON FOR TABLE only_null;
+
+# Inject statistics so that the current stats
+# cache is invalidated and creating partial
+# statistics has access to the latest full
+# statistic.
+statement ok
+ALTER TABLE only_null INJECT STATISTICS '$only_null_stat';
 
 statement error pq: only NULL values exist in the index, so partial stats cannot be collected
-CREATE STATISTICS only_null_partial ON a FROM a_null USING EXTREMES;
+CREATE STATISTICS only_null_partial ON a FROM only_null USING EXTREMES;
 
 statement ok
 CREATE INDEX ON xy (y) WHERE y > 5;

